File: server/java/src/main/java/com/zjournal/web/QueryEngine.java
package com.zjournal.web;

import org.springframework.util.MultiValueMap;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Reimplements the ad-hoc query engine from server/php/getters.php: repeated query keys
 * (e.g. ?id=a&id=b) OR their matches together, distinct keys AND across each other,
 * "field_like=value" does a substring match, and "_sort"/"_order" sort the final result.
 */
public final class QueryEngine {

    private static final String LIKE_SUFFIX = "_like";

    private QueryEngine() {
    }

    public static List<Map<String, Object>> apply(List<Map<String, Object>> items, MultiValueMap<String, String> params) {
        List<Map<String, Object>> result = items;
        String sortField = null;
        boolean descending = false;

        for (Map.Entry<String, List<String>> entry : params.entrySet()) {
            String key = entry.getKey();
            List<String> values = entry.getValue();
            if (values == null || values.isEmpty()) {
                continue;
            }

            if ("_sort".equals(key)) {
                sortField = values.get(0);
                continue;
            }
            if ("_order".equals(key)) {
                descending = values.get(0) != null && values.get(0).toLowerCase().contains("desc");
                continue;
            }

            boolean like = key.endsWith(LIKE_SUFFIX);
            String field = like ? key.substring(0, key.length() - LIKE_SUFFIX.length()) : key;

            Set<Map<String, Object>> seen = Collections.newSetFromMap(new IdentityHashMap<>());
            List<Map<String, Object>> matches = new ArrayList<>();
            for (String value : values) {
                for (Map<String, Object> item : result) {
                    if (matches(item, field, value, like) && seen.add(item)) {
                        matches.add(item);
                    }
                }
            }
            result = matches;
        }

        if (sortField != null) {
            String field = sortField;
            Comparator<Map<String, Object>> comparator = Comparator.comparing(item -> String.valueOf(item.get(field)));
            if (descending) {
                comparator = comparator.reversed();
            }
            result = result.stream().sorted(comparator).collect(Collectors.toList());
        }

        return result;
    }

    private static boolean matches(Map<String, Object> item, String field, String value, boolean like) {
        Object fieldValue = item.get(field);
        if (fieldValue == null) {
            return false;
        }
        String stringValue = String.valueOf(fieldValue);
        return like ? stringValue.contains(value) : stringValue.equalsIgnoreCase(value);
    }
}

File: server/python/app/query_engine.py
"""Reimplements com.zjournal.web.QueryEngine (Java), itself a reimplementation of the ad-hoc
query engine in server/php/getters.php: repeated query keys (e.g. ?id=a&id=b) OR their matches
together, distinct keys AND across each other, "field_like=value" does a substring match, and
"_sort"/"_order" sort the final result.
"""

from typing import Any

_LIKE_SUFFIX = "_like"


def apply(items: list[dict[str, Any]], params: list[tuple[str, str]]) -> list[dict[str, Any]]:
    result = items
    sort_field: str | None = None
    descending = False

    grouped: dict[str, list[str]] = {}
    order: list[str] = []
    for key, value in params:
        if key not in grouped:
            grouped[key] = []
            order.append(key)
        grouped[key].append(value)

    for key in order:
        values = grouped[key]
        if not values:
            continue

        if key == "_sort":
            sort_field = values[0]
            continue
        if key == "_order":
            descending = "desc" in values[0].lower()
            continue

        like = key.endswith(_LIKE_SUFFIX)
        field = key[: -len(_LIKE_SUFFIX)] if like else key

        matches = []
        seen_ids = set()
        for value in values:
            for item in result:
                if id(item) in seen_ids:
                    continue
                if _matches(item, field, value, like):
                    matches.append(item)
                    seen_ids.add(id(item))
        result = matches

    if sort_field is not None:
        result = sorted(result, key=lambda item: str(item.get(sort_field)), reverse=descending)

    return result


def _matches(item: dict[str, Any], field: str, value: str, like: bool) -> bool:
    field_value = item.get(field)
    if field_value is None:
        return False
    string_value = str(field_value)
    return value in string_value if like else string_value.lower() == value.lower()
